File: 0138-copy-list-with-random-pointer/0138-copy-list-with-random-pointer.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* next;
    Node* random;
    
    Node(int _val) {
        val = _val;
        next = NULL;
        random = NULL;
    }
};
*/

class Solution {
public:
    
   unordered_map<Node*, Node*> replacement;
    
    Node* copyRandomList(Node* head) {
        if(head == NULL) return NULL;
        
        if(replacement.find(head) != replacement.end()) return replacement[head];
        
        Node* newNode = new Node(head->val);
        
        replacement[head] = newNode;
        newNode->next = copyRandomList(head->next);
        newNode->random = copyRandomList(head->random);
        
        return replacement[head];
    }
};

File: 0138-copy-list-with-random-pointer/0138-copy-list-with-random-pointer.java
/*
// Definition for a Node.
class Node {
    int val;
    Node next;
    Node random;

    public Node(int val) {
        this.val = val;
        this.next = null;
        this.random = null;
    }
}
*/

class Solution {
    
    Map<Node, Node> replacement = new LinkedHashMap<Node, Node>();
    
    public Node copyRandomList(Node head) {
        
        if(head == null) return null;
        
        if(replacement.containsKey(head)) return replacement.get(head);
        
        Node newNode = new Node(head.val);
        
        replacement.put(head, newNode);
        newNode.next = copyRandomList(head.next);
        newNode.random = copyRandomList(head.random);
        
        return replacement.get(head);
    }  
        
}